src/utils/FixedQueue.ts

```diff
@@ -0,0 +1,52 @@
+import * as _ from 'underscore';
+import { deepClone } from '../../CloneUtils';
+
+export class FixedQueue<T> {
+
+  private queue: Array<T>;
+  private maxLength: number;
+
+  constructor(initialValues: Array<T> = [], maxLength: number = Number.MAX_VALUE) {
+    this.maxLength = maxLength;
+    this.queue = initialValues;
+    this.trimQueue();
+  }
+
+  immutablePush(object: any): FixedQueue<T> {
+    const queue: Array<T> = deepClone(this.queue);
+    queue.push(object);
```

Comment:
You are right, I somehow missed that the array was cloned as well. Still that is what I have been saying, a util class should simplify complex functions and what you are doing is complexifying simple function. It would be much cleaner and less error prone this way.